keys.go

```diff
@@ -133,22 +133,121 @@ type KeyVersion struct {
+	}
+}
+
+func WithPolicy(policy *Policy) CreateKeyOption {
+	return func(key *Key) string {
+		/*
+			Setting the value of rotationInterval to -1 in case user passes 0 value
+			as we want to retain the param `interval_month` after marshalling
+			so that we can get correct error msg from REST API saying interval_month should be between 1 to 12
+			Otherwise the param would not be sent to REST API in case of value 0
+			and it would throw error saying interval_month is missing
+		*/
+		if policy.Rotation != nil && policy.Rotation.Interval == 0 {
+			policy.Rotation.Interval = -1
+		}
+		key.Rotation = policy.Rotation
+		key.DualAuthDelete = policy.DualAuth
+		return "policy"
+		/*
+			I needed these CreateKeyOption functions to return a string to ensure I knew when a policy was passed,
+			so that it would call the create key with policy overrides function,
+			even when a policy with empty/nil fields is passed.
+		*/
+	}
+}
+
+func WithTags(tags []string) CreateKeyOption {
+	return func(key *Key) string {
+		key.Tags = tags
+		return "tags"
+	}
+}
+
+func (c *Client) CreateKeyWithOptions(ctx context.Context, name string, extractable bool, options ...CreateKeyOption) (*Key, error) {
+	key := &Key{
+		Name:        name,
+		Type:        keyType,
+		Extractable: extractable,
+	}
+	overrides := false
+	for _, opt := range options {
+		if opt != nil {
+			optName := opt(key)
+			if optName == "policy" {
+				overrides = true
+			}
+		}
+	}
+
+	path := keysPath
+	if overrides {
+		path = keysWithPolicyOverridesPath
+	}
+	return c.createKeyResource(ctx, *key, path)
+}
+
 // CreateKey creates a new KP key.
 func (c *Client) CreateKey(ctx context.Context, name string, expiration *time.Time, extractable bool) (*Key, error) {
-	return c.CreateImportedKey(ctx, name, expiration, "", "", "", extractable)
+	return c.CreateKeyWithOptions(ctx, name, extractable, WithExpiration(expiration))
 }
 
 // CreateImportedKey creates a new KP key from the given key material.
 func (c *Client) CreateImportedKey(ctx context.Context, name string, expiration *time.Time, payload, encryptedNonce, iv string, extractable bool) (*Key, error) {
-	key := c.createKeyTemplate(ctx, name, expiration, payload, encryptedNonce, iv, extractable, nil, AlgorithmRSAOAEP256, nil)
-	return c.createKey(ctx, key)
+	return c.CreateKeyWithOptions(ctx, name, extractable,
+		WithExpiration(expiration),
+		WithPayload(payload, encryptedNonce, iv, false),
+	)
 }
 
 // CreateImportedKeyWithSHA1 creates a new KP key from the given key material
 // using RSAES OAEP SHA 1 as encryption algorithm.
 func (c *Client) CreateImportedKeyWithSHA1(ctx context.Context, name string, expiration *time.Time, payload, encryptedNonce, iv string, extractable bool, aliases []string) (*Key, error) {
-	key := c.createKeyTemplate(ctx, name, expiration, payload, encryptedNonce, iv, extractable, aliases, AlgorithmRSAOAEP1, nil)
-	return c.createKey(ctx, key)
+	return c.CreateKeyWithOptions(ctx, name, extractable,
```

Comment:
Need to add `WithAliases()` for this case because it is a parameter in this function

Reply:
addressed.